.circleci/cull-old-ci-instances.py

```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
+
+# Runs periodically in it's own workflow in the CI/CD environment to teardown
+# instances that have exceeded a lifetime limit
+
+import datetime
+import pytz
+import boto3
+
+from common import get_instances_with_filter, unique_tag_key
+
+# The number of hours an instance may exist since its initial launch time
+INSTANCE_LIFETIME_LIMIT_HOURS = 24
+
+def main():
+    # Get a timezone-aware datetime instance
+    current_time = datetime.datetime.utcnow().replace(tzinfo=pytz.UTC)
+
+    # Grab all instances with a CI-generated tag
+    all_ci_instances_filter = {'Name': 'tag:' + unique_tag_key, 'Values' : ['*']}
+    all_ci_instances = get_instances_with_filter([all_ci_instances_filter], allowed_states=['*'])
+
+    client = boto3.client('ec2')
+    print "Terminated Instances:"
+    for inst in all_ci_instances:
+        lifetime_secs = (current_time - inst["LaunchTime"]).total_seconds()
```

Comment:
I'm glad I went and looked for more details on this.  Time since last state transition isn't a thing I guess.  I think I was remembering that you can cull instances that have been stopped for a period of time by parsing the datetime out of the transition reason string as in https://stackoverflow.com/questions/41231630/checking-stop-time-of-ec2-instance-with-boto3

Maybe not exactly what you're trying to do here but in case you find it useful there it is.

Reply:
Yeah I was aware of the string in the transition reason. It's something to consider doing in the future.